hathor/manager.py

```diff
@@ -550,58 +550,76 @@ def _verify_checkpoints(self) -> None:
-                assert len(other_genesis) == 1
-                return [self.tx_storage.get_tx(vertex_id) for vertex_id in other_genesis]
+                genesis_txs = [self._settings.GENESIS_TX1_HASH, self._settings.GENESIS_TX2_HASH]
+                if tx.is_transaction:
+                    other_genesis_tx, = set(genesis_txs) - {tx.hash}
+                    return [self.tx_storage.get_tx(other_genesis_tx)]
+                else:
+                    return [self.tx_storage.get_tx(t) for t in genesis_txs]
 
             parents = tx.get_tx_parents()
             assert len(parents) == 2
             return list(parents)
 
+        unconfirmed_tips = [tx for tx in self.tx_storage.iter_mempool_tips() if tx.timestamp < timestamp]
+        unconfirmed_extras = sorted(
+            (tx for tx in self.tx_storage.iter_mempool() if tx.timestamp < timestamp and tx not in unconfirmed_tips),
+            key=lambda tx: tx.timestamp,
+        )
+
+        # mix the blocks tx-parents, with their own tx-parents to avoid carrying one of the genesis tx over
+        best_block_tx_parents = get_tx_parents(best_block)
+        tx1_tx_grandparents = get_tx_parents(best_block_tx_parents[0])
+        tx2_tx_grandparents = get_tx_parents(best_block_tx_parents[1])
+        confirmed_tips = sorted(
+            set(best_block_tx_parents) | set(tx1_tx_grandparents) | set(tx2_tx_grandparents),
+            key=lambda tx: tx.timestamp,
+        )
+
         match unconfirmed_tips:
             case []:
-                if len(confirmed_tips) >= 2:
-                    return ParentTxs.from_txs(can_include=confirmed_tips, must_include=())
-                assert len(confirmed_tips) == 1
-                tx = confirmed_tips[0]
-                return ParentTxs.from_txs(can_include=get_tx_parents(tx), must_include=(tx,))
-
+                self.log.debug('generate_parent_txs: empty mempool, repeat parents')
+                return ParentTxs.from_txs(can_include=confirmed_tips[-2:], must_include=())
             case [tip_tx]:
-                if len(confirmed_tips) >= 1:
-                    return ParentTxs.from_txs(can_include=confirmed_tips, must_include=(tip_tx,))
-                return ParentTxs.from_txs(can_include=get_tx_parents(tip_tx), must_include=(tip_tx,))
-
+                if unconfirmed_extras:
+                    self.log.debug('generate_parent_txs: one tx tip and at least one other mempool tx')
+                    return ParentTxs.from_txs(can_include=unconfirmed_extras[-1:], must_include=(tip_tx,))
+                else:
+                    self.log.debug('generate_parent_txs: one tx in mempool, fill with one repeated parent')
+                    return ParentTxs.from_txs(can_include=confirmed_tips[-1:], must_include=(tip_tx,))
             case _:
+                self.log.debug('generate_parent_txs: multiple unconfirmed mempool tips')
                 return ParentTxs.from_txs(can_include=unconfirmed_tips, must_include=())
```

Comment:
I've kept the previous refactored structure as much as sensible, the main difference is that now `ParentTxs` will only have more than 2 transactions in the last case, when there are more than 2 unconfirmed tips, every other case should lead to `len(can_include) + len(must_include) == 2`, highest timestamp are prioritized when possible (sorting by timestamp and using `[-1:]` and `[-2:]`).